corehq/apps/api/user_updates.py

```diff
@@ -165,3 +167,70 @@ def _log_phone_number_change(new_phone_numbers, old_phone_numbers, user_change_l
 
     if change_messages:
         user_change_logger.add_change_message({'phone_numbers': change_messages})
+
+
+def _update_location(user, location_object, user_change_logger):
+    primary_location_id = location_object.get('primary_location')
+    location_ids = location_object.get('locations')
+
+    if primary_location_id is None and location_ids is None:
+        return
+
+    current_primary_location_id = user.get_location_id(user.domain)
+    current_locations = user.get_location_ids(user.domain)
+
+    if not primary_location_id and not location_ids:
+        _remove_all_locations(user, user_change_logger)
+    else:
+        if _validate_locations(primary_location_id, location_ids):
+            locations = _verify_location_ids(location_ids, user.domain)
+            if primary_location_id != current_primary_location_id:
+                _update_primary_location(user, primary_location_id, user_change_logger)
+            if set(current_locations) != set(location_ids):
+                _update_assigned_locations(user, locations, location_ids, user_change_logger)
+
+
+def _validate_locations(primary_location_id, location_ids):
+    if not primary_location_id and not location_ids:
+        return False
+    if not primary_location_id or not location_ids:
+        raise UpdateUserException(_('Both primary_location and locations must be provided together.'))
+    if primary_location_id not in location_ids:
+        raise UpdateUserException(_('Primary location must be included in the list of locations.'))
+    return True
+
+
+def _remove_all_locations(user, user_change_logger):
+    user.unset_location(commit=False)
+    user.reset_locations([], commit=False)
+    if user_change_logger:
+        user_change_logger.add_changes({'location_id': None})
+        user_change_logger.add_info(UserChangeMessage.primary_location_removed())
+        user_change_logger.add_changes({'assigned_location_ids': []})
+        user_change_logger.add_info(UserChangeMessage.assigned_locations_info([]))
+
+
+def _update_primary_location(user, primary_location_id, user_change_logger):
+    primary_location = SQLLocation.active_objects.get(location_id=primary_location_id)
+    user.set_location(primary_location, commit=False)
+    if user_change_logger:
+        user_change_logger.add_changes({'location_id': primary_location_id})
+        user_change_logger.add_info(UserChangeMessage.primary_location_info(primary_location))
+
+
+def _verify_location_ids(location_ids, domain):
+    locations = SQLLocation.active_objects.filter(location_id__in=location_ids, domain=domain)
+    _ = list(locations)  # forces evaluation and populates locations._result_cache which can be used later
+    real_ids = locations.location_ids()
```

Comment:
```suggestion
    real_ids = [loc.location_id for loc in locations]
```

Ok turns out I was still wrong and `locations.values_list(...)` will clone the query but not the result cache, therefore not using the cache. (`location_ids` calls `values_list`). So let's just do this in python instead.